### PR TITLE
Allow changing default storageclass names

### PR DESCRIPTION
**Pull request by Nikhil-Ladha**
Added a new field named `StorageClassName` to the CephFilesystems,
CephBlockPools, CephNonResilientPools, CephObjectStores, NFS and Encryption
spec that takes the user defined name of the storage class created for
the respective type of resource.
And, `VirtualizationStorageClassName` for SC created for ceph block pools in
virtualization environment.
Added `MaxLength`, `Pattern` and duplicate values validation on the field.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2112929

Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

### CODE REVIEW

**Comment by Nikhil-Ladha** [comment]
@umangachapagain @iamniting please review.

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ type StorageClusterSpec struct {
 	DefaultStorageProfile string `json:"defaultStorageProfile,omitempty"`
 
 	StorageProfiles []StorageProfile `json:"storageProfiles,omitempty"`
+
+	// +kubebuilder:validation:MaxLength=230
+	// StorageClassName is the string to be used as the prefix for
+	// the default storage classes created by ocs-operator.
+	StorageClassName string `json:"storageClassName,omitempty"`
```
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:112]: This should be StorageClassNamePrefix if you are considering it as a prefix.

File: api/v1/storagecluster_types.go [hunk 2]
```diff
@@ -105,6 +105,11 @@ type StorageClusterSpec struct {
 	DefaultStorageProfile string `json:"defaultStorageProfile,omitempty"`
 
 	StorageProfiles []StorageProfile `json:"storageProfiles,omitempty"`
+
+	// +kubebuilder:validation:MaxLength=230
```
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:109]: It's advisable to keep the value low so that you can create a storage class with a longer suffix in the future.

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:112]: If someone changes the prefix after creating the storage cluster, what will be the impact? Should we prevent such operations from happening in the first place?

Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:112]: Ack, will update the name

File: api/v1/storagecluster_types.go (hunk 2 above)
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:109]: Hmm ok, what should be the value in that case in your opinion? Something around 200?

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:112]: In that case, the operator will reconcile and create new storage classes with the updated names.
IMO, we should allow them to change the name if they want to, in that way they can delete older sc's if they are unused and create new one's.

**Review by iamniting — CHANGES_REQUESTED** [review]
@Nikhil-Ladha I don't think we are looking for a prefix solution. If I am not mistaken we should look for exact names. @umangachapagain WDYT?

**Comment by Nikhil-Ladha** [comment]
> @Nikhil-Ladha I don't think we are looking for a prefix solution. If I am not mistaken we should look for exact names. @umangachapagain WDYT?

Updated the code to change exact names of the sc's , PTAL.

File: api/v1/storagecluster_types.go [hunk 3]
```diff
@@ -105,6 +105,9 @@ type StorageClusterSpec struct {
 	DefaultStorageProfile string `json:"defaultStorageProfile,omitempty"`
 
 	StorageProfiles []StorageProfile `json:"storageProfiles,omitempty"`
+
+	// StorageClassNames contains the name of the storage classes created for the respective storage class types
+	StorageClassNames map[string]string `json:"storageClassNames,omitempty"`
```
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:110]: Cannot we add storageclass names to `ManagedResourcesSpec` Spec?

Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:110]: Good point, probably we can do that and keep all such details regarding auxiliary resources together.
WDYT @iamniting @umangachapagain ?

Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:110]: I gave another look at the API structure, and think this might not be the best option.
We have NFS SC's as well, that are not part of the `ManagedResourcesSpec` field, and the `ManagedResourcesSpec` also contains info about Ceph dashboard, toolbox and couple of other things. So, adding the `StorageClassNames` field to it might not sit well.

Comment (Madhu-1) [reply api/v1/storagecluster_types.go:110]: Just a thought ManagedResouceSpec contains different fields for block, file,nfs, etc you can add one more field storageclassname to reach specs to ManageCephBlockPools, ManageCephFilesystems, etc? and for NFS am not sure why it was kept outside of ManagedResouceSpec but you can add storageclassname to the NFS spec as well? 

@iamniting @umangachapagain WDYT?

Comment (umangachapagain) [reply api/v1/storagecluster_types.go:110]: I like the idea. Probably better in comparison to using map.

Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:110]: Ok, let me update the code accordingly.

**Comment by Nikhil-Ladha** [comment]
Updated the code to add the `StorageClassName` field to the `ManagedResources` properties.

File: api/v1/storagecluster_types.go [hunk 4]
```diff
@@ -206,6 +215,9 @@ type ManageCephObjectStores struct {
 	DisableStorageClass bool   `json:"disableStorageClass,omitempty"`
 	GatewayInstances    int32  `json:"gatewayInstances,omitempty"`
 	DisableRoute        bool   `json:"disableRoute,omitempty"`
+	// StorageClassName specifies the name of the storage class created for ceph obc's
+	// +kubebuilder:validation:MaxLength=253
+	StorageClassName string `json:"storageClassName,omitempty"`
```
[the remark this answers is not in the document]
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:220]: Yes, we need it for the `ceph-rgw` SC.

File: controllers/storagecluster/reconcile.go [hunk 5]
```diff
@@ -716,3 +727,48 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
...
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster) error {
+	scMap := make(map[string]string)
+	if sc.Spec.ManagedResources != (ocsv1.ManagedResourcesSpec{}) {
+		mr := sc.Spec.ManagedResources
+		if mr.CephBlockPools != (ocsv1.ManageCephBlockPools{}) && mr.CephBlockPools.StorageClassName != "" {
+			if _, has := scMap[mr.CephBlockPools.StorageClassName]; has {
+				return fmt.Errorf("duplicate storage class type provided")
```
[the remark this answers is not in the document]
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Updated.

File: api/v1/storagecluster_types.go [hunk 6]
```diff
@@ -184,11 +184,17 @@ type ManageCephBlockPools struct {
 	ReconcileStrategy    string `json:"reconcileStrategy,omitempty"`
 	DisableStorageClass  bool   `json:"disableStorageClass,omitempty"`
 	DisableSnapshotClass bool   `json:"disableSnapshotClass,omitempty"`
+	// StorageClassName specifies the name of the storage class created for ceph block pools
+	// +kubebuilder:validation:MaxLength=253
```
[the remark this answers is not in the document]
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:188]: That's the maximum length of string allowed for object names in Kubernetes, as per this [doc](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#:~:text=This%20means%20the%20name%20must,no%20more%20than%20253%20characters)

File: controllers/storagecluster/reconcile.go [hunk 7]
```diff
@@ -723,3 +734,48 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster) error {
+	scMap := make(map[string]string)
+	if sc.Spec.ManagedResources != (ocsv1.ManagedResourcesSpec{}) {
```
[the remark this answers is not in the document]
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go:740]: Ok, I think `ManagedResources` might never be null if there exists a storagecluster, but the other types like `CephNonResilientPools` and`CephObjectStores` could be null, right? Unless we explicitly enable/set some properties for them, isn't it?

File: api/v1/storagecluster_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iamniting) [reply api/v1/storagecluster_types.go:188]: I think this is not the same as the name, If I am not mistaken name's length is no more than 63 char. Can you pls verify that once in the cluster?

File: controllers/storagecluster/reconcile.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iamniting) [reply controllers/storagecluster/reconcile.go:740]: As this is not a pointer, we will be still able to access the variables, If they are not set they will be empty in case of a string.

File: api/v1/storagecluster_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:188]: Verified, it is same as name.
Also, I checked and we are not generating any other name using the storage class names. Probably, you can double check just to be double sure.

**Review by iamniting — CHANGES_REQUESTED** [review]
A new PR was merged today, Can you pls take care of this SC as well?

File: api/v1/storagecluster_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iamniting) [reply api/v1/storagecluster_types.go:188]: I think 100 char should be enough, @umangachapagain WDYT?

**Comment by Nikhil-Ladha** [comment]
> A new [PR](https://github.com/red-hat-storage/ocs-operator/pull/2076) was merged today, Can you pls take care of this SC as well?

Since, no new field was added for the kubevirt crd, how do you think we should take the SC name as the input from the user?
Maybe under the `CephBlockPools` struct itself named as `KubeVirtStorageClassName`? Since, it is copying the config of `CephBlockPools` and just adding few extra params to the SC created.

**Comment by Nikhil-Ladha** [comment]
> > > A new [PR](https://github.com/red-hat-storage/ocs-operator/pull/2076) was merged today, Can you pls take care of this SC as well?
> > 
> > 
> > Since, no new field was added for the kubevirt crd, how do you think we should take the SC name as the input from the user? Maybe under the `CephBlockPools` struct itself named as `KubeVirtStorageClassName`? Since, it is copying the config of `CephBlockPools` and just adding few extra params to the SC created.
> 
> I think so too

Done, added a new field named `KubeVirtStorageClassName` for the new SC.

File: api/v1/storagecluster_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (umangachapagain) [reply api/v1/storagecluster_types.go:188]: Why limit it to 100 if we can allow 253? Let's allow max possible length.

[the remark this answers is not in the document]
Comment (iamniting) [reply api/v1/storagecluster_types.go:188]: I am thinking of a case where we may use the storageclass name to generate the other names. So in that case storageclass name should be smaller than the allowed limit right?

File: controllers/storagecluster/reconcile.go [hunk 8]
```diff
@@ -723,3 +734,46 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster) error {
+	scMap := make(map[string]string)
+	if sc.Spec.ManagedResources.CephBlockPools.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephBlockPools.StorageClassName]; has {
+			return fmt.Errorf("duplicate storage class name provided")
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:742]: Error message needs more details. Which StorageClass?
Also, better to validate all names before returning error. Let's say I use same name for 5 storageclasses, it should return 1 error and not 4.

File: controllers/storagecluster/reconcile.go [hunk 9]
```diff
@@ -723,3 +734,46 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster) error {
+	scMap := make(map[string]string)
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:739]:
```suggestion
	scMap := make(map[string]bool)
```

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Review by umangachapagain — COMMENTED** [review]
1. What happens if I provide StorageClass names in an upgraded cluster? Or, if cluster already has default names and I provide custom names?
2. What happens if I change custom names later? 

Tests should cover these basic scenarios.

File: api/v1/storagecluster_types.go [hunk 10]
```diff
@@ -184,11 +184,21 @@ type ManageCephBlockPools struct {
...
+	StorageClassName string `json:"storageClassName,omitempty"`
+	// KubeVirtStorageClassName specifies the name of the storage class created for ceph block pools
+	// for virtualization environment
+	// +kubebuilder:validation:MaxLength=253
+	KubeVirtStorageClassName string `json:"kubeVirtStorageClassName,omitempty"`
```
Comment (malayparida2000) [inline api/v1/storagecluster_types.go:193]: Let's call it something like VirtualizationStorageClassName, As in the future we might create this storageclass for some other crd also apart from kubevirt.

File: controllers/storagecluster/reconcile.go (hunk 8 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Updated

File: controllers/storagecluster/reconcile.go (hunk 9 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Done

**Comment by Nikhil-Ladha** [comment]
> 1. What happens if I provide StorageClass names in an upgraded cluster? Or, if cluster already has default names and I provide custom names?
> 2. What happens if I change custom names later?
> 
> Tests should cover these basic scenarios.

Added these tests as well.

File: controllers/storagecluster/storageclasses.go [hunk 11]
```diff
@@ -97,7 +97,7 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 		sc := scc.storageClass
 
 		switch {
-		case strings.Contains(sc.Name, "-ceph-rbd") && !scc.isClusterExternal:
+		case (strings.Contains(sc.Name, "-ceph-rbd") || strings.Contains(sc.Provisioner, "rbd.csi.ceph.com")) && !scc.isClusterExternal:
```
Comment (umangachapagain) [inline controllers/storagecluster/storageclasses.go:100]: These changes don't look related to changing names.

File: api/v1/storagecluster_types.go [hunk 12]
```diff
@@ -184,11 +184,21 @@ type ManageCephBlockPools struct {
 	ReconcileStrategy    string `json:"reconcileStrategy,omitempty"`
 	DisableStorageClass  bool   `json:"disableStorageClass,omitempty"`
 	DisableSnapshotClass bool   `json:"disableSnapshotClass,omitempty"`
+	// StorageClassName specifies the name of the storage class created for ceph block pools
+	// +kubebuilder:validation:MaxLength=253
+	StorageClassName string `json:"storageClassName,omitempty"`
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:189]: We should add more validations to names. Kubernetes has multiple restrictions on naming. We should follow that to not hit into unwanted issue.
Ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

File: controllers/storagecluster/reconcile.go [hunk 13]
```diff
@@ -723,3 +734,57 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	var err error
+	scMap := make(map[string]bool)
+	if sc.Spec.ManagedResources.CephBlockPools.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephBlockPools.StorageClassName]; has {
+			err = fmt.Errorf("Duplicate storage class name provided for CephBlockPools")
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:743]: This error will be overridden by errors from other lines. We need to combine all errors and return it in the end.
And to reduce the amount of logs, let's not log it after every check. Let's log it once in the end.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/storageclasses.go (hunk 11 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go:100]: I did this change, because if we use custom names for SC then we would never hit the conditions in the switch the case i.e, to wait for the backend to be ready before creating the SC using them

File: controllers/storagecluster/reconcile.go (hunk 13 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go:743]: Use an array and append the errors to it and log all of them together?

File: api/v1/storagecluster_types.go (hunk 12 above)
Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:189]: In that case, we would have to use a pattern validation like this
```
// StorageClassName specifies the name of the storage class created for cephfs
// +kubebuilder:validation:MaxLength=253
// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
StorageClassName string `json:"storageClassName,omitempty"`
```
Ref: https://github.com/kubernetes/apimachinery/blob/v0.27.3/pkg/util/validation/validation.go#L215

Does this look fine?

Comment (Nikhil-Ladha) [reply api/v1/storagecluster_types.go:189]: Or, should we just use the `validation` pkg from `apimachinery` and validate the name in the `validateStorageClassname` function?

Comment (umangachapagain) [reply api/v1/storagecluster_types.go:189]: Pattern validation is good enough.

File: controllers/storagecluster/reconcile.go (hunk 13 above)
Comment (umangachapagain) [reply controllers/storagecluster/reconcile.go:743]: Yes. Can use errors.Join (needs go 1.20) to merge errors.

Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Ok, updated the code to use an array for concatenating all the errors and logging in the end of the function, as the project still uses go 1.19

File: controllers/storagecluster/reconcile.go [hunk 14]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	scMap := make(map[string]bool)
+	duplicateNames := []string{}
+	if sc.Spec.ManagedResources.CephBlockPools.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephBlockPools.StorageClassName]; has {
+			duplicateNames = append(duplicateNames, fmt.Sprintf("Duplicate storage class name provided for CephBlockPools: %s", sc.Spec.ManagedResources.CephBlockPools.StorageClassName))
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:743]:
```suggestion
			duplicateNames = append(duplicateNames, "CephBlockPools"))
```
Just store resources with duplicate storageclass names. Error message is already there in the end, so we can remove duplicates.

File: controllers/storagecluster/reconcile.go [hunk 15]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
...
+		}
+		scMap[sc.Spec.ManagedResources.CephBlockPools.StorageClassName] = true
+	}
+	if sc.Spec.ManagedResources.CephFilesystems.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephFilesystems.StorageClassName]; has {
+			duplicateNames = append(duplicateNames, fmt.Sprintf("Duplicate storage class name provided for CephFilesystems: %s", sc.Spec.ManagedResources.CephFilesystems.StorageClassName))
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:749]:
```suggestion
			duplicateNames = append(duplicateNames, "CephFilesystems"))
```

File: controllers/storagecluster/reconcile.go [hunk 16]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
...
+	}
+	if sc.Spec.Encryption.StorageClass && sc.Spec.Encryption.KeyManagementService.Enable && sc.Spec.Encryption.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.Encryption.StorageClassName]; has {
+			duplicateNames = append(duplicateNames, fmt.Sprintf("Duplicate storage class name provided for Encryption: %s", sc.Spec.Encryption.StorageClassName))
+		}
+		scMap[sc.Spec.Encryption.StorageClassName] = true
+	}
+
+	if len(duplicateNames) > 0 {
+		return fmt.Errorf("Duplicate storage class name(s) provided: %v", duplicateNames)
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:780]:
```suggestion
		return fmt.Errorf("Duplicate StorageClass name(s) provided for: %v", duplicateNames)
```

**Comment by Nikhil-Ladha** [comment]
Tested out the changes, and everything is working fine.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/reconcile.go (hunk 14 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Ahh ok, updated it now. Sorry for the trouble 😓

File: controllers/storagecluster/reconcile.go [hunk 17]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
...
+	}
+	if sc.Spec.ManagedResources.CephFilesystems.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephFilesystems.StorageClassName]; has {
+			duplicateNames = append(duplicateNames, "CephFilesystems")
+		}
+		scMap[sc.Spec.ManagedResources.CephFilesystems.StorageClassName] = true
+	}
+	if sc.Spec.ManagedResources.CephNonResilientPools.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephNonResilientPools.StorageClassName]; has {
+			duplicateNames = append(duplicateNames, "CephNonResilientPools: %s", sc.Spec.ManagedResources.CephNonResilientPools.StorageClassName)
```
Comment (umangachapagain) [inline controllers/storagecluster/reconcile.go:755]:
```suggestion
			duplicateNames = append(duplicateNames, "CephNonResilientPools")
```

**Review by umangachapagain — COMMENTED** [review]
/lgtm

**Comment by umangachapagain** [comment]
/hold
for other reviewers

File: controllers/storagecluster/reconcile.go [hunk 18]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
 	}
 	return nil
 }
+
+func validateCustomStorageClassNames(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	scMap := make(map[string]bool)
+	duplicateNames := []string{}
+	if sc.Spec.ManagedResources.CephBlockPools.StorageClassName != "" {
+		if _, has := scMap[sc.Spec.ManagedResources.CephBlockPools.StorageClassName]; has {
```
Comment (Madhu-1) [inline controllers/storagecluster/reconcile.go:742]: instead of `has` use `ok` as its a generic key used to check key exists.

File: controllers/storagecluster/storageclasses.go (hunk 11 above)
Comment (Madhu-1) [reply controllers/storagecluster/storageclasses.go:100]: instead of `rbd.csi.ceph.com` check for `<storagecluster-namespace>.rbd.csi-ceph.com`

File: controllers/storagecluster/storageclasses.go [hunk 19]
```diff
@@ -139,7 +139,7 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 				skippedSC = append(skippedSC, sc.Name)
 				continue
 			}
-		case strings.Contains(sc.Name, "-cephfs") && !scc.isClusterExternal:
+		case (strings.Contains(sc.Name, "-cephfs") || strings.Contains(sc.Provisioner, "cephfs.csi.ceph.com")) && !scc.isClusterExternal:
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:142]: instead of `cephfs.csi.ceph.com` check for `<storagecluster-namespace>.cephfs.csi-ceph.com`

File: controllers/storagecluster/storageclasses.go [hunk 20]
```diff
@@ -152,7 +152,7 @@ func (r *StorageClusterReconciler) createStorageClasses(sccs []StorageClassConfi
 				skippedSC = append(skippedSC, sc.Name)
 				continue
 			}
-		case strings.Contains(sc.Name, "-nfs"):
+		case strings.Contains(sc.Name, "-nfs") || strings.Contains(sc.Provisioner, "nfs.csi.ceph.com"):
```
Comment (Madhu-1) [inline controllers/storagecluster/storageclasses.go:155]: same here as well

**Review by Madhu-1 — COMMENTED** [review]
LGTM, small nit

File: controllers/storagecluster/storageclasses.go (hunk 19 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go:142]: Ok, for that I think we can use the `clusterID` from the SC for getting the namespace as it contains only the namespace of the storagecluster. Else, we will have to fetch it explicitly, wdyt @iamniting ?

Comment (Madhu-1) [reply controllers/storagecluster/storageclasses.go]: clusterID cannot be used even though it looks like a namespace name, but it can be any arbitrary name. My suggestion is to use a namespace where the Rook operator is running.

Comment (iamniting) [reply controllers/storagecluster/storageclasses.go]: @Madhu-1 What problem do you see if we just use `cephfs.csi.ceph.com`?

Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go]: Oh ok, you mean do something like this, right?
```
podList := &corev1.PodList{}
	err := r.Client.List(context.TODO(), podList, client.MatchingLabels{"app": "rook-ceph-operator"})
	if err != nil {
		r.Log.Error(err, "Failed to list rook-ceph-operator pod")
		return err
	}
	// Get the namespace in which the rook operator is running
	scNS := podList.Items[0].Namespace
```

Comment (Madhu-1) [reply controllers/storagecluster/storageclasses.go]: That can be used based on the assumption that no one is running any csi driver with that prefix for there CSI drivers.

Comment (Madhu-1) [reply controllers/storagecluster/storageclasses.go]: You can use the namespace of the storagecluster in your case or the ocs-operator namespace can also be used.

Comment (iamniting) [reply controllers/storagecluster/storageclasses.go]: We can get the namespace easily See this https://github.com/red-hat-storage/ocs-operator/blob/main/main.go#L121-L125

Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go]: Made the update to use the ocs-operator namespace

**Review by Madhu-1 — APPROVED** [review]
LGTM

File: controllers/storagecluster/storageclasses.go (hunk 19 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go]: Seems like the `OPERATOR_NAMESPACE` env variable is not set in the CI 🤔

Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go]: I think in that case we can add a double check for ns fetch, i.e, if the env variable is not there we can fetch the namespace from the `ocs-operator` namespace.

Comment (Nikhil-Ladha) [reply controllers/storagecluster/storageclasses.go]: Updated the test to set the `OPERATOR_NAMESPACE` env, looks good now.

**Comment by Nikhil-Ladha** [comment]
@iamniting can you please add your review as well.

File: controllers/storagecluster/reconcile.go [hunk 21]
```diff
@@ -723,3 +734,51 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
...
+	}
+	if sc.Spec.Encryption.StorageClass && sc.Spec.Encryption.KeyManagementService.Enable && sc.Spec.Encryption.StorageClassName != "" {
+		if _, ok := scMap[sc.Spec.Encryption.StorageClassName]; ok {
+			duplicateNames = append(duplicateNames, "Encryption")
+		}
+		scMap[sc.Spec.Encryption.StorageClassName] = true
+	}
+
+	if len(duplicateNames) > 0 {
+		return fmt.Errorf("Duplicate StorageClass name(s) provided for: %v", duplicateNames)
```
Comment (iamniting) [inline controllers/storagecluster/reconcile.go:780]:
```suggestion
		return fmt.Errorf("Duplicate StorageClass name(s) provided: %v", duplicateNames)
```

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/reconcile.go (hunk 21 above)
Comment (Nikhil-Ladha) [reply controllers/storagecluster/reconcile.go]: Done :)

**Review by iamniting — APPROVED** [review]

**Comment by Nikhil-Ladha** [comment]
Seems like the merge is not possible due to the `hold` label, I think we can remove it now @umangachapagain @iamniting
